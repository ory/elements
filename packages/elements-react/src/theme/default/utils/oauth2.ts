// Copyright Â© 2025 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import {
  OAuth2ConsentRequest,
  Session,
  UiContainer,
  UiNode,
  UiTextTypeEnum,
} from "@ory/client-fetch"
import { ConsentFlow } from "../../../util"

const rememberCheckbox: UiNode = {
  type: "input",
  group: "oauth2_consent",
  meta: {
    label: {
      id: 9999111,
      text: "Remember my decision",
      type: UiTextTypeEnum.Info,
    },
  },
  attributes: {
    node_type: "input",
    name: "remember",
    value: false,
    type: "checkbox",
    disabled: false,
  },
  messages: [],
}
const acceptButton: UiNode = {
  type: "input",
  group: "oauth2_consent",
  meta: {
    label: {
      id: 9999111,
      text: "Accept",
      type: UiTextTypeEnum.Info,
    },
  },
  attributes: {
    node_type: "input",
    name: "action",
    value: "accept",
    type: "submit",
    disabled: false,
  },
  messages: [],
}
const rejectButton: UiNode = {
  type: "input",
  group: "oauth2_consent",
  meta: {
    label: {
      id: 9999111,
      text: "Reject",
      type: UiTextTypeEnum.Info,
    },
  },
  attributes: {
    node_type: "input",
    name: "action",
    value: "reject",
    type: "submit",
    disabled: false,
  },
  messages: [],
}

export function translateConsentChallengeToUiNodes(
  consentChallenge: OAuth2ConsentRequest,
  csrfToken: string,
  formAction: string,
  session: Session,
): ConsentFlow {
  const ui: UiContainer = {
    action: formAction,
    nodes: [
      ...scopesToUiNodes(consentChallenge.requested_scope ?? []),
      rememberCheckbox,
      rejectButton,
      acceptButton,
      csrfTokenNode(csrfToken),
      challengeNode(consentChallenge.challenge),
    ],
    method: "POST",
    messages: [],
  }

  return {
    id: "UNSET",
    created_at: new Date(),
    expires_at: new Date(),
    issued_at: new Date(),
    state: "show_form",
    active: "oauth2_consent",
    ui,
    consent_request: consentChallenge,
    session,
  }
}

function scopesToUiNodes(scopes: string[]): UiNode[] {
  return scopes.map((scope) => ({
    type: "input",
    group: "oauth2_consent",
    meta: {
      label: {
        id: 9999111,
        text: scope,
        type: UiTextTypeEnum.Info,
      },
    },
    attributes: {
      node_type: "input",
      name: `grant_scope`,
      value: scope,
      type: "checkbox",
      disabled: false,
    },
    messages: [],
  }))
}

function csrfTokenNode(csrfToken: string): UiNode {
  return {
    type: "input",
    group: "default",
    meta: {},
    attributes: {
      node_type: "input",
      name: "csrf_token",
      value: csrfToken,
      type: "hidden",
      disabled: false,
    },
    messages: [],
  }
}

function challengeNode(challenge: string): UiNode {
  return {
    type: "input",
    group: "oauth2_consent",
    meta: {},
    attributes: {
      node_type: "input",
      name: "consent_challenge",
      value: challenge,
      type: "hidden",
      disabled: false,
    },
    messages: [],
  }
}
