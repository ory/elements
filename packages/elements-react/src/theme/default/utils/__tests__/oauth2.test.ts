// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { translateConsentChallengeToUiNodes } from "../oauth2"

describe("translate consent challenge to flow", () => {
  const testCases = {
    "no defined scope": { requested_scope: undefined },
    "empty scopes": { requested_scope: [] },
    "default scopes": { requested_scope: ["openid", "offline_access"] },
    "custom scopes": { requested_scope: ["custom_scope", "foo:bar"] },
  }

  const mockDate = new Date("2025-03-24T12:00:00Z")
  jest
    .spyOn(global, "Date")
    .mockImplementation(() => mockDate as unknown as Date)

  for (const [tc, { requested_scope }] of Object.entries(testCases)) {
    test(`- ${tc}`, () => {
      const consentFlow = translateConsentChallengeToUiNodes(
        { challenge: "consent-challenge", requested_scope },
        "csrfToken",
        "/foo/bar/oauth",
        { id: "session-id" },
      )
      expect(consentFlow).toMatchSnapshot()
    })
  }
})
