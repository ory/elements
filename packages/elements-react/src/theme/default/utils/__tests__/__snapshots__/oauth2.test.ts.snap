// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translate consent challenge to flow - custom scopes 1`] = `
{
  "active": "oauth2_consent",
  "consent_request": {
    "challenge": "consent-challenge",
    "requested_scope": [
      "custom_scope",
      "foo:bar",
    ],
  },
  "created_at": 2025-03-24T12:00:00.000Z,
  "expires_at": 2025-03-24T12:00:00.000Z,
  "id": "UNSET",
  "issued_at": 2025-03-24T12:00:00.000Z,
  "session": {
    "id": "session-id",
  },
  "state": "show_form",
  "ui": {
    "action": "/foo/bar/oauth",
    "messages": [],
    "method": "POST",
    "nodes": [
      {
        "attributes": {
          "disabled": false,
          "name": "grant_scope",
          "node_type": "input",
          "type": "checkbox",
          "value": "custom_scope",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "custom_scope",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "grant_scope",
          "node_type": "input",
          "type": "checkbox",
          "value": "foo:bar",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "foo:bar",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "remember",
          "node_type": "input",
          "type": "checkbox",
          "value": false,
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Remember my decision",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "action",
          "node_type": "input",
          "type": "submit",
          "value": "reject",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Reject",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "action",
          "node_type": "input",
          "type": "submit",
          "value": "accept",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Accept",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "csrf_token",
          "node_type": "input",
          "type": "hidden",
          "value": "csrfToken",
        },
        "group": "default",
        "messages": [],
        "meta": {},
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "consent_challenge",
          "node_type": "input",
          "type": "hidden",
          "value": "consent-challenge",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {},
        "type": "input",
      },
    ],
  },
}
`;

exports[`translate consent challenge to flow - default scopes 1`] = `
{
  "active": "oauth2_consent",
  "consent_request": {
    "challenge": "consent-challenge",
    "requested_scope": [
      "openid",
      "offline_access",
    ],
  },
  "created_at": 2025-03-24T12:00:00.000Z,
  "expires_at": 2025-03-24T12:00:00.000Z,
  "id": "UNSET",
  "issued_at": 2025-03-24T12:00:00.000Z,
  "session": {
    "id": "session-id",
  },
  "state": "show_form",
  "ui": {
    "action": "/foo/bar/oauth",
    "messages": [],
    "method": "POST",
    "nodes": [
      {
        "attributes": {
          "disabled": false,
          "name": "grant_scope",
          "node_type": "input",
          "type": "checkbox",
          "value": "openid",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "openid",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "grant_scope",
          "node_type": "input",
          "type": "checkbox",
          "value": "offline_access",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "offline_access",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "remember",
          "node_type": "input",
          "type": "checkbox",
          "value": false,
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Remember my decision",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "action",
          "node_type": "input",
          "type": "submit",
          "value": "reject",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Reject",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "action",
          "node_type": "input",
          "type": "submit",
          "value": "accept",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Accept",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "csrf_token",
          "node_type": "input",
          "type": "hidden",
          "value": "csrfToken",
        },
        "group": "default",
        "messages": [],
        "meta": {},
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "consent_challenge",
          "node_type": "input",
          "type": "hidden",
          "value": "consent-challenge",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {},
        "type": "input",
      },
    ],
  },
}
`;

exports[`translate consent challenge to flow - empty scopes 1`] = `
{
  "active": "oauth2_consent",
  "consent_request": {
    "challenge": "consent-challenge",
    "requested_scope": [],
  },
  "created_at": 2025-03-24T12:00:00.000Z,
  "expires_at": 2025-03-24T12:00:00.000Z,
  "id": "UNSET",
  "issued_at": 2025-03-24T12:00:00.000Z,
  "session": {
    "id": "session-id",
  },
  "state": "show_form",
  "ui": {
    "action": "/foo/bar/oauth",
    "messages": [],
    "method": "POST",
    "nodes": [
      {
        "attributes": {
          "disabled": false,
          "name": "remember",
          "node_type": "input",
          "type": "checkbox",
          "value": false,
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Remember my decision",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "action",
          "node_type": "input",
          "type": "submit",
          "value": "reject",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Reject",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "action",
          "node_type": "input",
          "type": "submit",
          "value": "accept",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Accept",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "csrf_token",
          "node_type": "input",
          "type": "hidden",
          "value": "csrfToken",
        },
        "group": "default",
        "messages": [],
        "meta": {},
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "consent_challenge",
          "node_type": "input",
          "type": "hidden",
          "value": "consent-challenge",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {},
        "type": "input",
      },
    ],
  },
}
`;

exports[`translate consent challenge to flow - no defined scope 1`] = `
{
  "active": "oauth2_consent",
  "consent_request": {
    "challenge": "consent-challenge",
    "requested_scope": undefined,
  },
  "created_at": 2025-03-24T12:00:00.000Z,
  "expires_at": 2025-03-24T12:00:00.000Z,
  "id": "UNSET",
  "issued_at": 2025-03-24T12:00:00.000Z,
  "session": {
    "id": "session-id",
  },
  "state": "show_form",
  "ui": {
    "action": "/foo/bar/oauth",
    "messages": [],
    "method": "POST",
    "nodes": [
      {
        "attributes": {
          "disabled": false,
          "name": "remember",
          "node_type": "input",
          "type": "checkbox",
          "value": false,
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Remember my decision",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "action",
          "node_type": "input",
          "type": "submit",
          "value": "reject",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Reject",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "action",
          "node_type": "input",
          "type": "submit",
          "value": "accept",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {
          "label": {
            "id": 9999111,
            "text": "Accept",
            "type": "info",
          },
        },
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "csrf_token",
          "node_type": "input",
          "type": "hidden",
          "value": "csrfToken",
        },
        "group": "default",
        "messages": [],
        "meta": {},
        "type": "input",
      },
      {
        "attributes": {
          "disabled": false,
          "name": "consent_challenge",
          "node_type": "input",
          "type": "hidden",
          "value": "consent-challenge",
        },
        "group": "oauth2_consent",
        "messages": [],
        "meta": {},
        "type": "input",
      },
    ],
  },
}
`;
